.gemspec

```diff
@@ -45,7 +45,10 @@ Gem::Specification.new do |s|
   # json_pure 2.0.2+ officially requires Ruby >= 2.0, but should have specified that in 2.0
   s.add_runtime_dependency(%q<json_pure>, "~> 1.8")
   # net-ssh is a runtime dependency of Puppet::Util::NetworkDevice::Transport::Ssh
-  s.add_runtime_dependency(%q<net-ssh>, "~> 2.1")
+  # Beaker 3.0.0 to 3.10.0 depends on net-ssh 3.3.0beta1
+  # Beaker 3.11.0+ depends on net-ssh 4.0+
+  # be lenient to allow module testing where Beaker and Puppet are in same Gemfile
+  s.add_runtime_dependency(%q<net-ssh>, [">= 3.0", "< 5"]) if Gem::Version.new(RUBY_VERSION) >= Gem::Version.new('2.0.0')
```

Comment:
Yup, I'm aware that `.gemspec` should not generally contain dynamic dependencies like this (one way to get around that is with `ext/mkrf_conf.rb`).  But note that this is not the main `.gemspec` for Puppet as a gem - that one is built during packaging by a rake task.  This `.gemspec` only influences a local `bundle install` (and is also shipped as part of agent packages which maintain fixed dependencies when built).  See https://tickets.puppetlabs.com/browse/PA-1077 for details on which files are used for what - it's a bit complicated currently.

I already tried to use the `[">= 2.1", "<5 "]` but unfortunately that still failed on Ruby 1.9.3.  You can see the commit at https://github.com/puppetlabs/puppet/commit/9e4cab2b34296b19b8ac5afc24bfff3bee3bf645 / CI result at https://travis-ci.org/puppetlabs/puppet/builds/221556267

I believe Bundler has a bug where it doesn't actually consider `RUBY_VERSION` when resolving dependencies without a version guard (they might have fixed it in newer versions, but I've definitely run into it before).  That *could* be addressed by adding even more dynamic resolution, but I think it's unnecessary (though that could be debated).  Another thing we can do is remove `net-ssh` altogether, but I believe that's misleading given there is code in Puppet that consumes `net-ssh`.